### DISCUSSION TITLE
Usage of ==

### DISCUSSION THREAD

**Discussion by pmknowles** [arendst/Tasmota#18699] [Q&A] [answered]
All of a sudden == is rearing its head.
When it was in expressions I (sort off) accepted it. Now it is being introduced into general rules
`on switch1#state==1`
These are memory constrained devices. Why introduce an extra character for no real benefit?
in C == is used to differentiate between assignment (x=1) and comparison (if x==1) but Tasmota doesn't use = for assignment.
The usage here is only for comparison with numbers. Text comparison is still using a single =. But Tasmota doesn't declare variables as strings or numbers.
In theory I could have 2 rules where in one I state var1 hello and another var1 1. Does that mean I need to have
`on var1#state=hello`
and in another
`on var1#state==1`?
Please nip this in the bud. We don't have reams and reams of code and == is just not needed.

**Comment by sfromis** [comment, marked as answer]
Using `==` in Tasmota rules for numeric comparisons has always been "the architecturally correct thing" to do, nothing has "suddenly" changed.

However, for most purposes using only integers (not variables with decimals), using `=` for string comparison also works for such values. Just forget about comparing to a programming language, rules is a **very** different thing from conventions about programming language constructs.

If your var1 is the result of a command like `Add1 1`, with the default `CalcRes 3`, the content would be `1.000` and a trigger condition like `var1#state=1` would always fail. Similar if doing `var1 01`. However, if your var1 is the result of a command like `var1 1`, or you are using `CalcRes 0`, the content would be just `1`, and both types of comparison works.

Nothing changed, except for docs updated for lower risk of leading you into confusion.....

Technically, the vars only contains strings, never a binary representation of the value. What you get by using `==` is to ensure that the compare parses both sides as numbers, instead of strings.
